### PR TITLE
Ceres: Add detailed documentation

### PR DESCRIPTION
**Pull request by iain-buclaw-sociomantic**
Based on existing documentation, some parts from `whisper.rst` where similarities crossed.  And some general experience.

What's immediately missing are:
- [ ] Configure carbon / graphite to write to and read from ceres. I can do a short section on this later.
- [ ] Differences between Ceres and Whisper, performance, disk space efficiency, etc.  It would be great if someone can chime in on this, as I've never really used Whisper.  But I could _guess_ from existing documentation on Whisper vs. RRD.

Initially putting this here so a proper review can be done, spelling, etc.

### CODE REVIEW

**Comment by iain-buclaw-sociomantic** [comment]
Of course, and if anything is not clear, put a big red flag on it.  I wrote many parts very quickly.

File: webapp/graphite/settings.py [hunk 1]
```diff
@@ -187,6 +187,12 @@
   except ImportError:
     print >> sys.stderr, "WARNING: whisper module could not be loaded, whisper support disabled"
   try:
+    import ceres  # noqa
+    if os.path.exists(CERES_DIR)
```
Comment (deniszh) [inline webapp/graphite/settings.py]: Missing :

File: docs/ceres.rst [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+Ceres is a time-series database format intended to replace Whisper as the default storage format
+for Graphite. In contrast with Whisper, Ceres is not a fixed-size database and is designed to
+better support sparse data of arbitrary fixed-size resolutions. This allows Graphite to distribute
+individual time-series across multiple servers or mounts.
+
+
+Storage Overview
+----------------
+Ceres databases are comprised of a single tree that stores all metrics in nesting directories as nodes,
+contained within a single path on disk.
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: Swap the second and third parts of this sentence around.

File: docs/ceres.rst [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+A Ceres node represents a single time-series metric, and is composed of at least two data files. A slice
+to store all data points, and an arbitrary key-value metadata file. The minimum required metadata a node
+needs is a ``'timeStep'``. This setting is the finest resolution that can be used for writing. A Ceres
+node however can contain and read data with other, less-precise values in its underlying slice data.
+
+Other metadata keys that may be set for compatibility with Graphite may be ``'retentions'`` ,
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: Double `may be`. Probably a rewording is needed.

File: docs/ceres.rst [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+``'xFilesFacter'``, and ``'aggregationMethod'``.
+
+A Ceres slice contains the actual data points in a file. The only other information a slice holds is the
+timestamp of the oldest data point stored, and the resolution. Both of which are encoded as part of its
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: s/stored//

File: docs/ceres.rst [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+the same way for special strings such as ``'inf'``. Maximum and minimum values are determined by the
+Python interpreter's allowable range for float values which can be found by executing::
+
+    python -c 'import sys; print sys.float_info'
+
+
+Slices: Precision and Fragmentation
+-----------------------------------
+Ceres databases contain one or more slices, each with a specific data resolution and a timestamp to mark
+the start time of the slice. Slices are ordered from the most recent timestamp to the oldest timestamp.
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: Maybe use beginning instead of start time?

File: docs/ceres.rst [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+Resolution of data is not considered when reading data, only that when writing a slice with the highest
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: Double `data`

File: docs/ceres.rst [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+precision configured for the node exists.
+
+Gaps in data are handled in Ceres by padding the data files with null datapoints. If the slice gap
+however is too big, then a new slice is instead created. If a Ceres node accumulates too many slices,
+performance can suffer. This can be caused by intermittently reported data. To mitigate slice
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: read performance

File: docs/ceres.rst [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+fragmentation there is a tolerance for how much space can be wasted within a slice file to avoid
+creating a new one. That tolerance level is determined by ``'MAX_SLICE_GAP'``, which is the number of
+consecutive null datapoints allowed in a slice file.
+
+If set very low, Ceres will waste less of the tiny bit disk space that this feature wastes, but then
+will be prone to performance problems caused by slice fragmentation, which can be pretty severe.
+
+If set really high, Ceres will waste a bit more disk space. Although each null datapoint wastes 8 bytes,
+you must keep in mind your filesystem's block size. If you suffer slice fragmentation issues, you should
```
Comment (iain-buclaw-sociomantic) [inline docs/ceres.rst]: Could say something about filesystem compression level (ZFS, Btrfs, Zram, ...).  Then again, maybe not.

Just thinking aloud. :-)

**Comment by obfuscurity** [comment]
@iain-buclaw-sociomantic Went ahead and merged this in so we have at least _something_ in there. It would be great to add those missing sections as your time permits.

**Comment by iain-buclaw-sociomantic** [comment]
@obfuscurity - Hi, OK, no problem.

I've just managed to get round to looking at this again, and got graphite-web set-up with some test ceres data.

> Configure carbon / graphite to write to and read from ceres. I can do a short section on this later.

Only thing I had to change on the configuration side in `local_settings.py` was:
```
STORAGE_FINDERS = (
    'graphite.finders.ceres.CeresFinder',
)
```
That's pretty much about it on the user side!  This configuration is already documented from the alternative storage finders page, so can check that off the list.

**Pull request merged**